File: repos/openjarvis/src/openjarvis/agents/monitor_operative.py
"""MonitorOperativeAgent -- long-horizon agent with configurable strategies.

Extends ToolUsingAgent (not OperativeAgent) with four configurable strategy
axes for long-horizon benchmark evaluation:

1. **memory_extraction** -- how findings are persisted to memory
2. **observation_compression** -- how tool outputs are compressed
3. **retrieval_strategy** -- how prior context is recalled
4. **task_decomposition** -- how complex tasks are split

The agent also inherits cross-session state persistence from the
OperativeAgent pattern (session_store, memory_backend, operator_id).
"""

from __future__ import annotations

import json
import logging
from typing import Any, List, Optional

from openjarvis.agents._stubs import AgentContext, AgentResult, ToolUsingAgent
from openjarvis.core.events import EventBus
from openjarvis.core.registry import AgentRegistry
from openjarvis.core.types import Message, Role, ToolCall, ToolResult
from openjarvis.engine._stubs import InferenceEngine
from openjarvis.tools._stubs import BaseTool

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Valid strategy values
# ---------------------------------------------------------------------------

VALID_MEMORY_EXTRACTION = {"causality_graph", "scratchpad", "structured_json", "none"}
VALID_OBSERVATION_COMPRESSION = {"summarize", "truncate", "none"}
VALID_RETRIEVAL_STRATEGY = {"hybrid_with_self_eval", "keyword", "semantic", "none"}
VALID_TASK_DECOMPOSITION = {"phased", "monolithic", "hierarchical"}

# ---------------------------------------------------------------------------
# Default system prompt
# ---------------------------------------------------------------------------

MONITOR_OPERATIVE_SYSTEM_PROMPT = """\
You are a Monitor Operative Agent designed for long-horizon tasks.

## Capabilities
1. TOOLS: Call any available tool via function calling
2. STATE: Your previous findings and state are automatically restored
3. MEMORY: Store important findings for future recall

## Strategy
- Memory extraction: {memory_extraction}
- Observation compression: {observation_compression}
- Retrieval strategy: {retrieval_strategy}
- Task decomposition: {task_decomposition}

## Protocol
- Break complex tasks into phases and track progress
- Store causal relationships and key findings in memory
- Compress long tool outputs before adding to context
- Self-evaluate retrieved context for relevance
- Always persist state before finishing

{tool_descriptions}"""


@AgentRegistry.register("monitor_operative")
class MonitorOperativeAgent(ToolUsingAgent):
    """Long-horizon agent with configurable memory, compression, retrieval,
    and decomposition strategies.

    The four strategy axes control how the agent manages information across
    turns and sessions:

    - ``memory_extraction``: How findings are persisted (causality_graph,
      scratchpad, structured_json, none).
    - ``observation_compression``: How tool outputs are compressed before
      being added to context (summarize, truncate, none).
    - ``retrieval_strategy``: How prior context is recalled at the start
      of each run (hybrid_with_self_eval, keyword, semantic, none).
    - ``task_decomposition``: How complex tasks are broken down
      (phased, monolithic, hierarchical).
    """

    agent_id = "monitor_operative"
    accepts_tools = True

    def __init__(
        self,
        engine: InferenceEngine,
        model: str,
        *,
        tools: Optional[List[BaseTool]] = None,
        bus: Optional[EventBus] = None,
        max_turns: int = 25,
        temperature: float = 0.3,
        max_tokens: int = 4096,
        system_prompt: Optional[str] = None,
        # Strategy parameters
        memory_extraction: str = "causality_graph",
        observation_compression: str = "summarize",
        retrieval_strategy: str = "hybrid_with_self_eval",
        task_decomposition: str = "phased",
        # State persistence (OperativeAgent pattern)
        operator_id: Optional[str] = None,
        session_store: Optional[Any] = None,
        memory_backend: Optional[Any] = None,
        **kwargs: Any,
    ) -> None:
        super().__init__(
            engine, model, tools=tools, bus=bus,
            max_turns=max_turns, temperature=temperature,
            max_tokens=max_tokens,
        )
        # Validate strategies
        if memory_extraction not in VALID_MEMORY_EXTRACTION:
            raise ValueError(
                f"Invalid memory_extraction {memory_extraction!r}, "
                f"must be one of {VALID_MEMORY_EXTRACTION}"
            )
        if observation_compression not in VALID_OBSERVATION_COMPRESSION:
            raise ValueError(
                f"Invalid observation_compression {observation_compression!r}, "
                f"must be one of {VALID_OBSERVATION_COMPRESSION}"
            )
        if retrieval_strategy not in VALID_RETRIEVAL_STRATEGY:
            raise ValueError(
                f"Invalid retrieval_strategy {retrieval_strategy!r}, "
                f"must be one of {VALID_RETRIEVAL_STRATEGY}"
            )
        if task_decomposition not in VALID_TASK_DECOMPOSITION:
            raise ValueError(
                f"Invalid task_decomposition {task_decomposition!r}, "
                f"must be one of {VALID_TASK_DECOMPOSITION}"
            )

        self._memory_extraction = memory_extraction
        self._observation_compression = observation_compression
        self._retrieval_strategy = retrieval_strategy
        self._task_decomposition = task_decomposition

        self._system_prompt = system_prompt
        self._operator_id = operator_id
        self._session_store = session_store
        self._memory_backend = memory_backend

    # ------------------------------------------------------------------
    # Main run loop
    # ------------------------------------------------------------------

    def run(
        self,
        input: str,
        context: Optional[AgentContext] = None,
        **kwargs: Any,
    ) -> AgentResult:
        """Execute the agent on *input* with the configured strategies."""
        self._emit_turn_start(input)

        # 1. Build system prompt with state context
        sys_parts: list[str] = []
        if self._system_prompt:
            sys_parts.append(self._system_prompt)
        else:
            tool_desc = self._build_tool_descriptions()
            try:
                sys_parts.append(
                    MONITOR_OPERATIVE_SYSTEM_PROMPT.format(
                        memory_extraction=self._memory_extraction,
                        observation_compression=self._observation_compression,
                        retrieval_strategy=self._retrieval_strategy,
                        task_decomposition=self._task_decomposition,
                        tool_descriptions=tool_desc,
                    ),
                )
            except KeyError:
                sys_parts.append(MONITOR_OPERATIVE_SYSTEM_PROMPT)

        # 2. State recall from memory backend
        previous_state = self._recall_state()
        if previous_state:
            sys_parts.append(f"\n## Previous State\n{previous_state}")

        system_prompt = "\n\n".join(sys_parts) if sys_parts else None

        # 3. Load session history
        session_messages = self._load_session()

        # 4. Build messages
        messages = self._build_operative_messages(
            input, context,
            system_prompt=system_prompt,
            session_messages=session_messages,
        )

        # 5. Run function-calling tool loop
        openai_tools = self._executor.get_openai_tools() if self._tools else []
        all_tool_results: list[ToolResult] = []
        turns = 0
        content = ""
        state_stored_by_tool = False

        for _turn in range(self._max_turns):
            turns += 1

            gen_kwargs: dict[str, Any] = {}
            if openai_tools:
                gen_kwargs["tools"] = openai_tools

            result = self._generate(messages, **gen_kwargs)
            content = result.get("content", "")
            raw_tool_calls = result.get("tool_calls", [])

            # No tool calls -> check continuation, then final answer
            if not raw_tool_calls:
                content = self._check_continuation(result, messages)
                break

            # Build ToolCall objects from raw dicts
            tool_calls = [
                ToolCall(
                    id=tc.get("id", f"call_{i}"),
                    name=tc.get("name", ""),
                    arguments=tc.get("arguments", "{}"),
                )
                for i, tc in enumerate(raw_tool_calls)
            ]

            # Append assistant message with tool calls
            messages.append(Message(
                role=Role.ASSISTANT,
                content=content,
                tool_calls=tool_calls,
            ))

            # Execute each tool
            for tc in tool_calls:
                # Loop guard check
                if self._loop_guard:
                    verdict = self._loop_guard.check_call(
                        tc.name, tc.arguments,
                    )
                    if verdict.blocked:
                        tool_result = ToolResult(
                            tool_name=tc.name,
                            content=f"Loop guard: {verdict.reason}",
                            success=False,
                        )
                        all_tool_results.append(tool_result)
                        messages.append(Message(
                            role=Role.TOOL,
                            content=tool_result.content,
                            tool_call_id=tc.id,
                            name=tc.name,
                        ))
                        continue

                tool_result = self._executor.execute(tc)
                all_tool_results.append(tool_result)

                # Track explicit state storage
                if tc.name == "memory_store" and self._operator_id:
                    try:
                        args = json.loads(tc.arguments)
                        state_key = f"monitor_operative:{self._operator_id}:state"
                        if args.get("key", "") == state_key:
                            state_stored_by_tool = True
                    except (json.JSONDecodeError, TypeError) as exc:
                        logger.debug(
                            "Failed to parse tool call arguments"
                            " for state tracking: %s", exc,
                        )

                # Compress observation if strategy requires it
                observation_content = self._compress_observation(tool_result.content)

                messages.append(Message(
                    role=Role.TOOL,
                    content=observation_content,
                    tool_call_id=tc.id,
                    name=tc.name,
                ))

                # Extract and store findings based on memory strategy
                self._extract_and_store(tc.name, tool_result.content)
        else:
            # Max turns exceeded
            self._save_session(input, content)
            return self._max_turns_result(
                all_tool_results, turns, content=content,
            )

        # 6. Save session
        self._save_session(input, content)

        # 7. Auto-persist state if agent didn't do it explicitly
        if not state_stored_by_tool:
            self._auto_persist_state(content)

        self._emit_turn_end(turns=turns, content_length=len(content))
        return AgentResult(
            content=content,
            tool_results=all_tool_results,
            turns=turns,
        )

    # ------------------------------------------------------------------
    # Message building
    # ------------------------------------------------------------------

    def _build_operative_messages(
        self,
        input: str,
        context: Optional[AgentContext],
        *,
        system_prompt: Optional[str] = None,
        session_messages: Optional[list[Message]] = None,
    ) -> list[Message]:
        """Build message list with system prompt, session history, and input."""
        messages: list[Message] = []
        if system_prompt:
            messages.append(Message(role=Role.SYSTEM, content=system_prompt))
        if session_messages:
            messages.extend(session_messages)
        if context and context.conversation.messages:
            messages.extend(context.conversation.messages)
        messages.append(Message(role=Role.USER, content=input))
        return messages

    def _build_tool_descriptions(self) -> str:
        """Build a text description of available tools for the system prompt."""
        if not self._tools:
            return ""
        from openjarvis.tools._stubs import build_tool_descriptions
        return build_tool_descriptions(self._tools)

    # ------------------------------------------------------------------
    # Strategy methods
    # ------------------------------------------------------------------

    def _compress_observation(self, content: str) -> str:
        """Compress a tool observation according to the compression strategy.

        - ``summarize``: If content exceeds 2000 chars, ask the LLM to
          summarize. Falls back to truncation if generation fails.
        - ``truncate``: Hard-truncate at 2000 chars with an ellipsis.
        - ``none``: Return content unchanged.
        """
        if self._observation_compression == "none":
            return content
        if self._observation_compression == "truncate":
            if len(content) > 2000:
                return content[:2000] + "\n... [truncated]"
            return content
        # "summarize"
        if len(content) <= 2000:
            return content
        try:
            summary_messages = [
                Message(
                    role=Role.SYSTEM,
                    content="Summarize the following tool output concisely, "
                    "preserving all key facts and data points.",
                ),
                Message(role=Role.USER, content=content[:8000]),
            ]
            result = self._generate(summary_messages)
            summary = result.get("content", "")
            if summary:
                return summary
        except Exception:
            logger.debug("Observation summarization failed, falling back to truncation")
        # Fallback to truncation
        return content[:2000] + "\n... [truncated]"

    def _extract_and_store(self, tool_name: str, content: str) -> None:
        """Extract findings from a tool result and store them.

        The extraction strategy depends on ``_memory_extraction``:

        - ``causality_graph``: Extract causal relationships via
          :meth:`_extract_causality` and store as KG triples.
        - ``scratchpad``: Append raw content to a scratchpad key in
          memory.
        - ``structured_json``: Attempt to parse JSON from the content
          and store structured data.
        - ``none``: Do nothing.
        """
        if self._memory_extraction == "none":
            return
        if not self._memory_backend:
            return

        if self._memory_extraction == "causality_graph":
            self._extract_causality(tool_name, content)
        elif self._memory_extraction == "scratchpad":
            self._store_scratchpad(tool_name, content)
        elif self._memory_extraction == "structured_json":
            self._store_structured(tool_name, content)

    def _extract_causality(self, tool_name: str, content: str) -> None:
        """Extract causal relationships from tool output and store them.

        Uses the LLM to identify cause-effect patterns, then stores
        them via the memory backend.
        """
        if not self._memory_backend or not content.strip():
            return
        # Only attempt extraction for substantial outputs
        if len(content) < 50:
            return
        try:
            extract_messages = [
                Message(
                    role=Role.SYSTEM,
                    content=(
                        "Extract causal relationships from the following tool "
                        "output. Return a JSON array of objects with 'cause', "
                        "'effect', and 'confidence' fields. If no causal "
                        "relationships are found, return an empty array []."
                    ),
                ),
                Message(role=Role.USER, content=content[:4000]),
            ]
            result = self._generate(extract_messages)
            raw = result.get("content", "")
            # Try to parse JSON from the response
            raw = raw.strip()
            if raw.startswith("```"):
                # Strip code fences
                lines = raw.split("\n")
                raw = "\n".join(lines[1:-1] if len(lines) > 2 else lines)
            relations = json.loads(raw)
            if isinstance(relations, list):
                operator_prefix = (
                    f"monitor_operative:{self._operator_id}"
                    if self._operator_id
                    else "monitor_operative"
                )
                for rel in relations[:10]:  # Cap at 10 per extraction
                    if isinstance(rel, dict) and "cause" in rel and "effect" in rel:
                        key = f"{operator_prefix}:causality:{rel['cause'][:50]}"
                        value = json.dumps(rel)
                        try:
                            self._memory_backend.store(key, value)
                        except Exception as exc:
                            logger.debug(
                                "Failed to store causality relation in memory: %s", exc,
                            )
        except (json.JSONDecodeError, Exception):
            logger.debug(
                "Causality extraction failed for tool %s output", tool_name,
            )

    def _store_scratchpad(self, tool_name: str, content: str) -> None:
        """Append content to a scratchpad entry in memory."""
        if not self._memory_backend:
            return
        operator_prefix = (
            f"monitor_operative:{self._operator_id}"
            if self._operator_id
            else "monitor_operative"
        )
        key = f"{operator_prefix}:scratchpad:{tool_name}"
        # Truncate long content
        snippet = content[:1000] if len(content) > 1000 else content
        try:
            self._memory_backend.store(key, snippet)
        except Exception:
            logger.debug("Could not store scratchpad for tool %s", tool_name)

    def _store_structured(self, tool_name: str, content: str) -> None:
        """Try to parse JSON from tool output and store structured data."""
        if not self._memory_backend:
            return
        operator_prefix = (
            f"monitor_operative:{self._operator_id}"
            if self._operator_id
            else "monitor_operative"
        )
        try:
            data = json.loads(content)
            key = f"{operator_prefix}:structured:{tool_name}"
            self._memory_backend.store(key, json.dumps(data))
        except (json.JSONDecodeError, TypeError):
            # Not JSON -- store as plain text truncated
            key = f"{operator_prefix}:structured:{tool_name}"
            try:
                self._memory_backend.store(key, content[:1000])
            except Exception as exc:
                logger.debug(
                    "Failed to store structured data for tool %s: %s",
                    tool_name, exc,
                )

    # ------------------------------------------------------------------
    # State persistence (OperativeAgent pattern)
    # ------------------------------------------------------------------

    def _recall_state(self) -> str:
        """Retrieve previous state from memory backend."""
        if not self._memory_backend or not self._operator_id:
            return ""
        state_key = f"monitor_operative:{self._operator_id}:state"
        try:
            result = self._memory_backend.retrieve(state_key)
            if result:
                return result if isinstance(result, str) else str(result)
        except Exception:
            logger.debug(
                "No previous state for monitor_operative %s",
                self._operator_id,
            )
        return ""

    def _load_session(self) -> list[Message]:
        """Load recent session history for this operator."""
        if not self._session_store or not self._operator_id:
            return []
        session_id = f"monitor_operative:{self._operator_id}"
        try:
            session = self._session_store.get_or_create(session_id)
            if hasattr(session, "messages") and session.messages:
                recent = session.messages[-10:]
                return [
                    Message(
                        role=Role(m.get("role", "user")),
                        content=m.get("content", ""),
                    )
                    for m in recent
                    if isinstance(m, dict)
                ]
        except Exception:
            logger.debug(
                "Could not load session for monitor_operative %s",
                self._operator_id,
            )
        return []

    def _save_session(self, input_text: str, response: str) -> None:
        """Save the tick's prompt and response to the session store."""
        if not self._session_store or not self._operator_id:
            return
        session_id = f"monitor_operative:{self._operator_id}"
        try:
            self._session_store.save_message(
                session_id, {"role": "user", "content": input_text},
            )
            self._session_store.save_message(
                session_id, {"role": "assistant", "content": response},
            )
        except Exception:
            logger.debug(
                "Could not save session for monitor_operative %s",
                self._operator_id,
            )

    def _auto_persist_state(self, content: str) -> None:
        """Auto-persist a state summary if agent didn't store explicitly."""
        if not self._memory_backend or not self._operator_id:
            return
        state_key = f"monitor_operative:{self._operator_id}:state"
        try:
            summary = content[:1000] if content else ""
            self._memory_backend.store(state_key, summary)
        except Exception:
            logger.debug(
                "Could not auto-persist state for monitor_operative %s",
                self._operator_id,
            )


__all__ = ["MonitorOperativeAgent", "MONITOR_OPERATIVE_SYSTEM_PROMPT"]

File: repos/openjarvis/rust/crates/openjarvis-agents/src/monitor_operative.rs
//! MonitorOperativeAgent -- long-horizon monitoring agent with configurable strategies.
//!
//! Implements the monitor-operative pattern with four strategy axes:
//! 1. Memory extraction (extract key info from observations)
//! 2. Observation compression (compress verbose outputs)
//! 3. Retrieval (search memory for relevant context)
//! 4. Task decomposition (break complex tasks into subtasks)

use crate::loop_guard::LoopGuard;
use crate::traits::OjAgent;
use crate::utils::strip_think_tags;
use openjarvis_core::{AgentContext, AgentResult, OpenJarvisError, ToolResult};
use openjarvis_tools::executor::ToolExecutor;
use regex::Regex;
use rig::agent::AgentBuilder;
use rig::completion::message::Message as RigMessage;
use rig::completion::request::{Chat, CompletionModel};
use std::collections::HashMap;
use std::sync::Arc;

// ---------------------------------------------------------------------------
// Strategy enums
// ---------------------------------------------------------------------------

/// How findings are persisted to memory.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MemoryExtraction {
    /// Extract causal relationships via LLM and store as structured entries.
    CausalityGraph,
    /// Append raw content to a scratchpad key.
    Scratchpad,
    /// Attempt to parse JSON from tool output and store structured data.
    StructuredJson,
    /// Do not extract or store anything.
    None,
}

/// How tool outputs are compressed before adding to context.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ObservationCompression {
    /// Ask the LLM to summarize long outputs.
    Summarize,
    /// Hard-truncate at a character limit.
    Truncate,
    /// Return content unchanged.
    None,
}

/// How prior context is recalled at the start of each run.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RetrievalStrategy {
    /// Hybrid retrieval with self-evaluation of relevance.
    HybridWithSelfEval,
    /// Keyword-based retrieval.
    Keyword,
    /// Semantic similarity retrieval.
    Semantic,
    /// No retrieval.
    None,
}

/// How complex tasks are broken down.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TaskDecomposition {
    /// Break tasks into sequential phases.
    Phased,
    /// Execute as a single monolithic task.
    Monolithic,
    /// Hierarchical decomposition into subtask tree.
    Hierarchical,
}

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

/// Configuration for the monitor-operative agent's four strategy axes.
#[derive(Debug, Clone)]
pub struct MonitorConfig {
    pub memory_extraction: MemoryExtraction,
    pub observation_compression: ObservationCompression,
    pub retrieval_strategy: RetrievalStrategy,
    pub task_decomposition: TaskDecomposition,
    /// Maximum characters before compression kicks in.
    pub compression_threshold: usize,
    /// Maximum characters for truncation.
    pub truncation_limit: usize,
}

impl Default for MonitorConfig {
    fn default() -> Self {
        Self {
            memory_extraction: MemoryExtraction::CausalityGraph,
            observation_compression: ObservationCompression::Summarize,
            retrieval_strategy: RetrievalStrategy::HybridWithSelfEval,
            task_decomposition: TaskDecomposition::Phased,
            compression_threshold: 2000,
            truncation_limit: 2000,
        }
    }
}

// ---------------------------------------------------------------------------
// System prompt
// ---------------------------------------------------------------------------

fn build_system_prompt(config: &MonitorConfig, tool_list: &str) -> String {
    format!(
        "You are a Monitor Operative Agent designed for long-horizon tasks.\n\n\
         ## Capabilities\n\
         1. TOOLS: Call any available tool via function calling\n\
         2. STATE: Your previous findings and state are automatically restored\n\
         3. MEMORY: Store important findings for future recall\n\n\
         ## Strategy\n\
         - Memory extraction: {memory_extraction:?}\n\
         - Observation compression: {observation_compression:?}\n\
         - Retrieval strategy: {retrieval_strategy:?}\n\
         - Task decomposition: {task_decomposition:?}\n\n\
         ## Protocol\n\
         - Break complex tasks into phases and track progress\n\
         - Store causal relationships and key findings in memory\n\
         - Compress long tool outputs before adding to context\n\
         - Self-evaluate retrieved context for relevance\n\
         - Always persist state before finishing\n\n\
         Available tools: {tool_list}\n\n\
         For each step, output:\n\
         Thought: <your reasoning>\n\
         Action: <tool_name>\n\
         Action Input: <JSON arguments>\n\n\
         After receiving an observation, continue reasoning.\n\
         When you have the final answer, output:\n\
         Thought: I now know the answer.\n\
         Final Answer: <your answer>",
        memory_extraction = config.memory_extraction,
        observation_compression = config.observation_compression,
        retrieval_strategy = config.retrieval_strategy,
        task_decomposition = config.task_decomposition,
        tool_list = tool_list,
    )
}

// ---------------------------------------------------------------------------
// Agent implementation
// ---------------------------------------------------------------------------

/// Long-horizon monitoring agent with configurable memory, compression,
/// retrieval, and decomposition strategies.
///
/// Uses a multi-turn Thought-Action-Observation loop (similar to
/// `NativeReActAgent`) augmented with strategy-driven observation
/// compression, memory extraction, and task decomposition.
pub struct MonitorOperativeAgent<M: CompletionModel> {
    agent: rig::agent::Agent<M>,
    executor: Arc<ToolExecutor>,
    max_turns: usize,
    config: MonitorConfig,
}

impl<M: CompletionModel> MonitorOperativeAgent<M> {
    pub fn new(
        model: M,
        executor: Arc<ToolExecutor>,
        max_turns: usize,
        temperature: f64,
        config: MonitorConfig,
    ) -> Self {
        let tool_list = executor.list_tools().join(", ");
        let system_prompt = build_system_prompt(&config, &tool_list);

        let agent = AgentBuilder::new(model)
            .preamble(&system_prompt)
            .temperature(temperature)
            .build();

        Self {
            agent,
            executor,
            max_turns,
            config,
        }
    }

    /// Parse `Action:` and `Action Input:` lines from model output.
    fn parse_action(text: &str) -> Option<(String, String)> {
        let action_re = Regex::new(r"(?m)^Action:\s*(.+)$").unwrap();
        let input_re = Regex::new(r"(?m)^Action Input:\s*(.+)$").unwrap();

        let action = action_re
            .captures(text)?
            .get(1)?
            .as_str()
            .trim()
            .to_string();
        let input = input_re
            .captures(text)
            .and_then(|c| c.get(1))
            .map(|m| m.as_str().trim().to_string())
            .unwrap_or_else(|| "{}".to_string());

        Some((action, input))
    }

    /// Parse `Final Answer:` from model output.
    fn parse_final_answer(text: &str) -> Option<String> {
        let re = Regex::new(r"(?m)^Final Answer:\s*(.+)").unwrap();
        re.captures(text)
            .and_then(|c| c.get(1))
            .map(|m| m.as_str().trim().to_string())
    }

    /// Compress an observation according to the configured strategy.
    fn compress_observation(&self, content: &str) -> String {
        match self.config.observation_compression {
            ObservationCompression::None => content.to_string(),
            ObservationCompression::Truncate => {
                if content.len() > self.config.truncation_limit {
                    let mut truncated = content[..self.config.truncation_limit].to_string();
                    truncated.push_str("\n... [truncated]");
                    truncated
                } else {
                    content.to_string()
                }
            }
            ObservationCompression::Summarize => {
                // For summarization we would call the LLM, but since we only
                // have the rig agent (chat interface) and not a raw model
                // handle, we fall back to truncation in the Rust implementation.
                // A production build could issue a side-channel generate call.
                if content.len() > self.config.compression_threshold {
                    let mut truncated = content[..self.config.truncation_limit].to_string();
                    truncated.push_str("\n... [summarized/truncated]");
                    truncated
                } else {
                    content.to_string()
                }
            }
        }
    }

    /// Build metadata reflecting the active strategy configuration.
    fn strategy_metadata(&self) -> HashMap<String, serde_json::Value> {
        let mut meta = HashMap::new();
        meta.insert(
            "memory_extraction".to_string(),
            serde_json::Value::String(format!("{:?}", self.config.memory_extraction)),
        );
        meta.insert(
            "observation_compression".to_string(),
            serde_json::Value::String(format!("{:?}", self.config.observation_compression)),
        );
        meta.insert(
            "retrieval_strategy".to_string(),
            serde_json::Value::String(format!("{:?}", self.config.retrieval_strategy)),
        );
        meta.insert(
            "task_decomposition".to_string(),
            serde_json::Value::String(format!("{:?}", self.config.task_decomposition)),
        );
        meta
    }

    /// Decompose input into subtask prompts according to the task decomposition
    /// strategy.  For `Monolithic` the original input is returned as-is.
    /// For `Phased` and `Hierarchical` the input is wrapped with decomposition
    /// instructions so the LLM itself performs the breakdown.
    fn decompose_input(&self, input: &str) -> String {
        match self.config.task_decomposition {
            TaskDecomposition::Monolithic => input.to_string(),
            TaskDecomposition::Phased => {
                format!(
                    "Break the following task into sequential phases and execute them one at a time.\n\
                     Task: {input}"
                )
            }
            TaskDecomposition::Hierarchical => {
                format!(
                    "Decompose the following task into a hierarchy of subtasks, then execute from leaves to root.\n\
                     Task: {input}"
                )
            }
        }
    }
}

#[async_trait::async_trait]
impl<M: CompletionModel + 'static> OjAgent for MonitorOperativeAgent<M> {
    fn agent_id(&self) -> &str {
        "monitor_operative"
    }

    fn accepts_tools(&self) -> bool {
        true
    }

    async fn run(
        &self,
        input: &str,
        context: Option<&AgentContext>,
    ) -> Result<AgentResult, OpenJarvisError> {
        let mut history: Vec<RigMessage> = context
            .map(|ctx| {
                ctx.conversation
                    .messages
                    .iter()
                    .filter_map(|m| match m.role {
                        openjarvis_core::Role::User => {
                            Some(RigMessage::user(&m.content))
                        }
                        openjarvis_core::Role::Assistant => {
                            Some(RigMessage::assistant(&m.content))
                        }
                        _ => None,
                    })
                    .collect()
            })
            .unwrap_or_default();

        let mut all_tool_results: Vec<ToolResult> = Vec::new();
        let mut guard = LoopGuard::default();

        // Apply task decomposition strategy to the input.
        let decomposed_input = self.decompose_input(input);
        let mut current_input = decomposed_input;

        for turn in 1..=self.max_turns {
            let response = self
                .agent
                .chat(&current_input, history.clone())
                .await
                .map_err(|e| {
                    OpenJarvisError::Agent(openjarvis_core::error::AgentError::Execution(
                        e.to_string(),
                    ))
                })?;

            let text = strip_think_tags(&response);

            // Check for final answer
            if let Some(answer) = Self::parse_final_answer(&text) {
                let mut metadata = self.strategy_metadata();
                metadata.insert(
                    "turns_used".to_string(),
                    serde_json::Value::Number(serde_json::Number::from(turn)),
                );
                return Ok(AgentResult {
                    content: answer,
                    tool_results: all_tool_results,
                    turns: turn,
                    metadata,
                });
            }

            // Check for action (tool call)
            if let Some((action, action_input)) = Self::parse_action(&text) {
                // Loop guard check
                if let Some(loop_msg) = guard.check(&action, &action_input) {
                    return Ok(AgentResult {
                        content: format!("Agent stopped: {}", loop_msg),
                        tool_results: all_tool_results,
                        turns: turn,
                        metadata: self.strategy_metadata(),
                    });
                }

                let params: serde_json::Value =
                    serde_json::from_str(&action_input).unwrap_or(serde_json::json!({}));

                let tool_result = match self.executor.execute(
                    &action,
                    &params,
                    Some("monitor_operative"),
                    None,
                ) {
                    Ok(r) => r,
                    Err(e) => ToolResult::failure(&action, e.to_string()),
                };

                // Compress observation according to strategy
                let compressed = self.compress_observation(&tool_result.content);

                history.push(RigMessage::assistant(&text));
                current_input = format!("Observation: {}", compressed);

                all_tool_results.push(tool_result);
            } else {
                // No action and no final answer -- treat as final response
                return Ok(AgentResult {
                    content: text,
                    tool_results: all_tool_results,
                    turns: turn,
                    metadata: self.strategy_metadata(),
                });
            }
        }

        // Max turns exceeded
        let mut metadata = self.strategy_metadata();
        metadata.insert(
            "max_turns_exceeded".to_string(),
            serde_json::Value::Bool(true),
        );
        Ok(AgentResult {
            content: format!("Reached maximum turns ({})", self.max_turns),
            tool_results: all_tool_results,
            turns: self.max_turns,
            metadata,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use openjarvis_engine::rig_adapter::RigModelAdapter;
    type MonitorAgent = MonitorOperativeAgent<RigModelAdapter<openjarvis_engine::Engine>>;

    #[test]
    fn test_parse_action() {
        let text =
            "Thought: I need to search\nAction: web_search\nAction Input: {\"query\": \"rust\"}";
        let (action, input) = MonitorAgent::parse_action(text).unwrap();
        assert_eq!(action, "web_search");
        assert!(input.contains("rust"));
    }

    #[test]
    fn test_parse_final_answer() {
        let text = "Thought: I know the answer\nFinal Answer: The result is 42.";
        let answer = MonitorAgent::parse_final_answer(text).unwrap();
        assert_eq!(answer, "The result is 42.");
    }

    #[test]
    fn test_compress_observation_none() {
        let config = MonitorConfig {
            observation_compression: ObservationCompression::None,
            ..Default::default()
        };
        // We can test compress_observation without constructing the full agent
        // by checking the strategy logic directly.
        assert_eq!(config.observation_compression, ObservationCompression::None);
    }

    #[test]
    fn test_default_config() {
        let config = MonitorConfig::default();
        assert_eq!(config.memory_extraction, MemoryExtraction::CausalityGraph);
        assert_eq!(
            config.observation_compression,
            ObservationCompression::Summarize
        );
        assert_eq!(
            config.retrieval_strategy,
            RetrievalStrategy::HybridWithSelfEval
        );
        assert_eq!(config.task_decomposition, TaskDecomposition::Phased);
        assert_eq!(config.compression_threshold, 2000);
        assert_eq!(config.truncation_limit, 2000);
    }

    #[test]
    fn test_strategy_enum_debug() {
        // Ensure Debug formatting works (used in system prompt and metadata).
        assert_eq!(format!("{:?}", MemoryExtraction::CausalityGraph), "CausalityGraph");
        assert_eq!(format!("{:?}", ObservationCompression::Truncate), "Truncate");
        assert_eq!(
            format!("{:?}", RetrievalStrategy::HybridWithSelfEval),
            "HybridWithSelfEval"
        );
        assert_eq!(format!("{:?}", TaskDecomposition::Hierarchical), "Hierarchical");
    }
}
